File: docs/legacy(old PACE)/C#/PaceAPISample/sample/GangedActivityJobCost.cs
using efipaceservices;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Web.Services.Protocols;

namespace Pace_Web_Service_SDK.sample
{
    class GangedActivityJobCost
    {
        private static ReadObjectPortType readObjectHttpBinding = PaceClient.getReadObjectHttpBinding();
        private static InvokeActionPortType invokeActionHttpBinding = PaceClient.getInvokeActionHttpBinding();

        public static void Run()
        {
            Console.WriteLine("Enter the employee id");
            String employeeCode = Console.ReadLine();

            Employee employee = SignInEmployee(employeeCode);

            List<JobPart> jobPartsToGang = new List<JobPart>();

            string customerCode = "HOUSE";
            int orderQty = 12345;

            jobPartsToGang.Add(CreateJob(customerCode, orderQty));
            jobPartsToGang.Add(CreateJob(customerCode, orderQty));
            jobPartsToGang.Add(CreateJob(customerCode, orderQty));

            string activityCode = "20420";

            ActivityCode activity = GetActivityCode(activityCode);

            /*
                 *
                 * Start a ganged job transaction
                 *
                 * @param rpcEmployee  Employee
                 * @param jobParts  List of Job Parts to create job transactions for
                 * @param activityCode ActivityCode
                 * @param hours  Hours
                 * @param prodUnits  Production Units
                 * @param complete Compelte
                 * @param beginCount  Begin Count
                 * @param endCount End Count
                 * @param beginMeter Begin Meter
                 * @param endMeter End Meter
                 * @param notes Notes
                 * @param nonPlannedReason NonPlannedResonID
                 * @return JobCost's created
                 * @throws XFireFault
                 */

            // Elapse time example

            List<JobCost> gangedActivities = invokeActionHttpBinding.startGangedJobTransaction(employee,
                                                                jobPartsToGang.ToArray(),
                                                                activity,
                                                                0,
                                                                0,
                                                                false,
                                                                0,
                                                                0,
                                                                0,
                                                                0,
                                                                null,
                                                                null).ToList();

            Console.WriteLine("Started Ganged Job Activity");

            System.Threading.Thread.Sleep(1000 * 60 * 2);

            invokeActionHttpBinding.pauseGangedJobTransaction(employee, gangedActivities.ToArray());

            Console.WriteLine("Paused Ganged Job Activity");

            System.Threading.Thread.Sleep(1000 * 60 * 2);

            invokeActionHttpBinding.resumeGangedJobTransaction(employee, gangedActivities.ToArray());

            Console.WriteLine("Resumed Ganged Job Activity");

            System.Threading.Thread.Sleep(1000 * 60 * 2);

            /*
                 *
                 * Complete a ganged job transaction
                 *
                 * @param rpcEmployee  Employee
                 * @param gangedCosts  Existing list of ganged JobCosts
                 * @param prodUnits        Production Units
                 * @param complete         Compelte
                 * @param beginCount       Begin Count
                 * @param endCount         End Count
                 * @param beginMeter       Begin Meter
                 * @param endMeter         End Meter
                 * @param notes            Notes
                 * @param nonPlannedReason NonPlannedResonID
                 *
                 * @return JobCost's created
                 */

            gangedActivities = invokeActionHttpBinding.completeGangedJobTransaction(employee,
                                                                                        gangedActivities.ToArray(),
                                                                                        0,
                                                                                        true,
                                                                                        1,
                                                                                        99,
                                                                                        99,
                                                                                        50,
                                                                                        "Ganged Activity Note",
                                                                                        null).ToList();

            Console.WriteLine("Completed Ganged Job Activity");

        }

        private static ActivityCode GetActivityCode( string activityCode)
        {
            ActivityCode activity = new ActivityCode();
            activity.id = activityCode;
            activity = readObjectHttpBinding.readActivityCode(activity);

            return activity;            
        }

        private static Employee SignInEmployee(string employeeCode)
        {
            UpdateObjectPortType updateObjectHttpBinding = PaceClient.getUpdateObjectHttpBinding();            

            Employee employee = new Employee();
            employee.id = employeeCode;
            employee = readObjectHttpBinding.readEmployee(employee);

            invokeActionHttpBinding.employeeSignIn(employee);

            Console.WriteLine("Signed in Employee " + employee.firstName + " " + employee.lastName);

            return employee;
        }

        private static JobPart CreateJob(string customerCode, int orderQty)
        {
            CreateJob createJob = new CreateJob();
            Job job = createJob.CreateJobForCustomer(customerCode, orderQty);

            return getJobPart( job, "01");
        }

        private static JobPart getJobPart(Job job1, string part)
        {
            JobPart jobPart = new JobPart();
            jobPart.job = job1.job;
            jobPart.jobPart = part;
            return readObjectHttpBinding.readJobPart(jobPart);
        }
    }
}

File: docs/legacy(old PACE)/java/src/main/com/pace2020/epace/sdk/sample/GangedActivityJobCost.java
/*
 * Copyright (c) 2017, Electronics for Imaging, Inc. EFI-Pace All Rights Reserved.
 */

package com.pace2020.epace.sdk.sample;

import java.math.BigDecimal;
import java.rmi.RemoteException;

import com.pace2020.epace.object.ActivityCode;
import com.pace2020.epace.object.ArrayOfJobCost;
import com.pace2020.epace.object.ArrayOfJobPart;
import com.pace2020.epace.object.Employee;
import com.pace2020.epace.object.JobPart;
import com.pace2020.epace.sdk.util.SecuredWSDLClient;

/**
 * @author <a href="mailto:jerry.duval@efi.com">jerry duval</a>
 */
public class GangedActivityJobCost extends SecuredWSDLClient
{
    public static void main( final String[] args ) throws Exception
    {
        new GangedActivityJobCost().run();
    }

    public void run() throws RemoteException, InterruptedException
    {
        final String employeeCode = "AA";

        Employee employee = signInEmployee( employeeCode );

        final ArrayOfJobPart jobPartsToGang = new ArrayOfJobPart();

        final String customerCode = "HOUSE";
        final String orderQty = "12345";

        jobPartsToGang.getJobPart().add( createJob( customerCode, orderQty ) );
        jobPartsToGang.getJobPart().add( createJob( customerCode, orderQty ) );
        jobPartsToGang.getJobPart().add( createJob( customerCode, orderQty ) );

        final String activityCode = "20420";

        final ActivityCode activity = getActivityCode( activityCode );

        /**
             *
             * Start a ganged job transaction
             *
             * @param rpcEmployee  Employee
             * @param jobParts  List of Job Parts to create job transactions for
             * @param activityCode ActivityCode
             * @param hours  Hours
             * @param prodUnits  Production Units
             * @param complete Compelte
             * @param beginCount  Begin Count
             * @param endCount End Count
             * @param beginMeter Begin Meter
             * @param endMeter End Meter
             * @param notes Notes
             * @param nonPlannedReason NonPlannedResonID
             * @return JobCost's created
             * @throws XFireFault
             */

        // Elapse time example

        ArrayOfJobCost gangedActivities = getInvokeActionPortType().startGangedJobTransaction( employee,
                                                                                               jobPartsToGang,
                                                                                               activity,
                                                                                               new Double( "0" ),
                                                                                               new Double( "0" ),
                                                                                               false,
                                                                                               new Double( "0" ),
                                                                                               new Double( "0" ),
                                                                                               new BigDecimal( "0" ),
                                                                                               new BigDecimal( "0" ),
                                                                                               null,
                                                                                               null );

        System.out.println( "Started Ganged Job Activity" );

        Thread.sleep( 1000 * 60 * 2 );  // 1000 milli * 60 seconds = 1 min    testing elapsed time

        getInvokeActionPortType().pauseGangedJobTransaction( employee, gangedActivities );

        System.out.println( "Paused Ganged Job Activity" );

        Thread.sleep( 1000 * 60 * 2 );  // 1000 milli * 60 seconds = 1 min    testing elapsed time

        getInvokeActionPortType().resumeGangedJobTransaction( employee, gangedActivities );

        System.out.println( "Resumed Ganged Job Activity" );

        Thread.sleep( 1000 * 60 * 2 );  // 1000 milli * 60 seconds = 1 min    testing elapsed time

        /**
             *
             * Complete a ganged job transaction
             *
             * @param rpcEmployee  Employee
             * @param gangedCosts  Existing list of ganged JobCosts
             * @param prodUnits        Production Units
             * @param complete         Compelte
             * @param beginCount       Begin Count
             * @param endCount         End Count
             * @param beginMeter       Begin Meter
             * @param endMeter         End Meter
             * @param notes            Notes
             * @param nonPlannedReason NonPlannedResonID
             *
             * @return JobCost's created
             */

        gangedActivities =
            getInvokeActionPortType().completeGangedJobTransaction( employee,
                                                                    gangedActivities,
                                                                    new Double( "0" ),
                                                                    true,
                                                                    new Double( "1" ),
                                                                    new Double( "99" ),
                                                                    new BigDecimal( "99" ),
                                                                    new BigDecimal( "50" ),
                                                                    "Ganged Activity Note",
                                                                    null );

        System.out.println( "Completed Ganged Job Activity" );
    }

    private ActivityCode getActivityCode( final String activityCode )
    {
        ActivityCode activity = new ActivityCode();

        activity.setId( activityCode );

        activity = getReadObjectPortType().readActivityCode( activity );
        return activity;
    }

    private Employee signInEmployee( final String employeeCode )
    {
        Employee employee = new Employee();

        employee.setId( employeeCode );

        employee = getReadObjectPortType().readEmployee( employee );

        getInvokeActionPortType().employeeSignIn( employee );

        System.out.println( "Signed in Employee " + employee.getFirstName() + " " + employee.getLastName() );

        return employee;
    }

    private JobPart createJob( final String customerCode, final String orderQty ) throws RemoteException
    {
        final CreateJob createJob = new CreateJob( customerCode, orderQty );

        return getJobPart( "01", createJob.run() );
    }
}